File: content/snippets/cpp/design_patterns/builder.cpp
#include <iostream>
#include <string>
#include <utility>
#include <vector>

struct Report {
    std::string title;
    std::vector<std::string> sections;
};

class ReportBuilder {
public:
    ReportBuilder& title(std::string value) {
        report_.title = std::move(value);
        return *this;
    }

    ReportBuilder& section(std::string value) {
        report_.sections.push_back(std::move(value));
        return *this;
    }

    Report build() {
        return report_;
    }

private:
    Report report_;
};

int main() {
    Report report = ReportBuilder()
        .title("Daily Build")
        .section("tests: green")
        .section("deploy: staged")
        .build();

    std::cout << report.title << '\n';
    for (const auto& section : report.sections) {
        std::cout << "- " << section << '\n';
    }
}

File: content/snippets/python/design_patterns/builder.py
from dataclasses import dataclass, field


@dataclass
class Report:
    title: str = ""
    sections: list[str] = field(default_factory=list)


class ReportBuilder:
    def __init__(self) -> None:
        self._report = Report()

    def title(self, value: str) -> "ReportBuilder":
        self._report.title = value
        return self

    def section(self, value: str) -> "ReportBuilder":
        self._report.sections.append(value)
        return self

    def build(self) -> Report:
        return self._report


report = (
    ReportBuilder()
    .title("Daily Build")
    .section("tests: green")
    .section("deploy: staged")
    .build()
)

print(report.title)
for section in report.sections:
    print(f"- {section}")